### ISSUE TITLE
Feature Request: Dynamic In-Loop Evaluation ("Fluid Eval" Callback)

### ISSUE THREAD

**Issue by ada-ggf25** [allenai/OLMo-core#491]
## Summary

I propose implementing a **Dynamic In-Loop Evaluation Callback** that periodically evaluates the model during training on adaptively-selected tasks, using principles from Fluid LM Benchmarking. Instead of static evaluation after training, this callback would test the model on appropriate tasks that adapt to its current ability level, providing real-time insight into training progress and enabling informed decisions about early stopping or curriculum adjustments.

## Motivation

Current in-loop evaluation in OLMo-core (via `EvaluatorCallback` and `DownstreamEvaluatorCallbackConfig`) uses fixed task sets that don't adapt to the model's evolving capabilities. This has several limitations:

1. **Inefficient evaluation**: Early in training, models are tested on tasks that are too difficult, providing little signal. Later, they're tested on tasks that are too easy, wasting compute.

2. **Evaluation noise**: Fixed evaluation sets can include items that are too easy or too unstable, leading to misleading comparisons and unreliable metrics (Heineman et al., 2025; see [Academic Background](#academic-background) below).

3. **Limited training insights**: Static evaluation doesn't provide dynamic feedback that could inform curriculum learning or adaptive training strategies.

4. **Research opportunity**: Recent work on Fluid LM Benchmarking (Hofmann et al., 2025; see [Academic Background](#academic-background) below) demonstrates that adaptive evaluation can achieve 50× fewer items on MMLU with lower variance, using item-response theory to dynamically select questions.

### Core Components

1. **FluidEvaluatorCallback**
   - Extends the existing `EvaluatorCallback` infrastructure
   - Works with evaluators backed by [OLMES](https://github.com/allenai/olmes) for task execution, but does not require OLMES as a dependency
   - Implements adaptive task selection based on model's current performance

2. **Adaptive Task Selection**
   - Consumes difficulty metadata for tasks/questions when available (for example from Fluid / [OLMES](https://github.com/allenai/olmes)), and falls back to heuristic difficulty signals (recent accuracy/confidence) otherwise
   - Selects questions that yield informative signal for the model's current performance level
   - Progressively includes harder tasks as model accuracy improves (similar to adaptive testing)
   - Can start with easier subtasks and adaptively increase difficulty

3. **Dynamic Evaluation Mix**
   - Adjusts the evaluation task mix based on prior in-loop results
   - Mimics Fluid Benchmark's adaptive item selection to reduce variance and increase validity
   - Can subsample tasks adaptively (e.g., 1-2 representative tasks per capability at each interval)

4. **Integration Points**
   - Extends `olmo_core.train.callbacks.EvaluatorCallback`
   - Can interoperate with the [OLMES](https://github.com/allenai/olmes) Python API for task execution when desired (optional)
   - Can run asynchronously to avoid stalling training (background process or validation cluster)
   - For minimal implementation, can start with perplexity on held-out validation set as a proxy

### Example API

```python
@dataclass
class FluidEvaluatorCallbackConfig(CallbackConfig):
    """Configuration for adaptive in-loop evaluation using fluid benchmarking principles."""
    
    # Optional OLMES integration (via OLMES-backed evaluators)
    evaluators: List[EvaluatorConfig]  # e.g., evaluators wrapping OLMES task sets or benchmark configs
    
    # Adaptive selection parameters
    adaptive_selection: bool = True
    """Whether to use adaptive task/question selection."""
    
    difficulty_estimation: Literal["confidence", "irt", "accuracy"] = "confidence"
    """Method for estimating question difficulty (IRT used in Phase 2)."""
    
    min_items_per_eval: int = 10
    """Minimum number of items to evaluate on."""
    
    max_items_per_eval: int = 100
    """Maximum number of items to evaluate on."""
    
    target_information: float | None = None
    """Optional target information gain per evaluation (Phase 2: IRT-based selection)."""
    
    # Evaluation schedule
    eval_interval: int = 1000
    """Steps between evaluations."""
    
    eval_on_startup: bool = False
    """Run evaluation at training start."""
    
    # Performance optimization
    async_eval: bool = False
    """Run evaluation asynchronously to not block training."""
    
    subsample_tasks: bool = True
    """Subsample tasks adaptively rather than running full suite."""
    
    tasks_per_capability: int = 1
    """Number of representative tasks per capability domain."""
```

### Implementation Approach

At a high level, the implementation would introduce:

1. A `FluidEvaluatorCallback` class that subclasses `EvaluatorCallback` and owns the adaptive selection logic while reusing the existing eval loop and metric logging.
2. A small per-task/item statistics buffer on the callback (e.g., rolling accuracy or confidence, plus optional difficulty metadata when available from Fluid / [OLMES](https://github.com/allenai/olmes)).
3. An adaptive sampler component or helper that, given the stats buffer and config (target accuracy band, max items per eval, etc.), chooses which evaluators/items to run on each eval step.
4. Configuration via `FluidEvaluatorCallbackConfig` so this can be plugged into existing trainer configs alongside `LMEvaluatorCallbackConfig` and `DownstreamEvaluatorCallbackConfig`.

## Academic Background

This proposal is grounded in recent research on adaptive evaluation:

1. **Fluid LM Benchmarking: Adapting Language Model Evaluation to Each Model** (Hofmann et al., 2025)
   - ArXiv: [2509.11106](https://arxiv.org/abs/2509.11106)
   - Demonstrates that adapting evaluation to each LM significantly boosts efficiency and reliability
   - Achieves 50× fewer items on MMLU with lower variance using item-response theory (IRT) for dynamic question selection

2. **Signal and Noise: A Framework for Reducing Uncertainty in Language Model Evaluation** (Heineman et al., 2025)
   - ArXiv: [2508.13144](https://arxiv.org/abs/2508.13144)
   - Introduces signal and noise metrics to analyze benchmark reliability
   - Shows that filtering out too-easy or too-unstable eval items improves reliability
   - Demonstrates that benchmarks with better signal-to-noise ratio are more reliable for decision-making

### Phase 1: Basic Adaptive Selection (MVP)

- Implement `FluidEvaluatorCallback` extending `EvaluatorCallback`
- Add simple difficulty estimation (e.g., based on model confidence or historical accuracy)
- Implement basic adaptive item selection (select items near model's estimated ability)
- Optionally integrate with [OLMES](https://github.com/allenai/olmes) for task execution (when using OLMES-backed evaluators)
- Add unit tests

### Phase 2: IRT-Based Selection

- Implement IRT-based difficulty estimation
- Add item bank with difficulty parameters
- Implement information-gain-based item selection
- Add configuration for IRT parameters

### Phase 3: Advanced Features

- Asynchronous evaluation execution
- Task subsampling and capability-based selection
- Integration with curriculum learning callbacks
- Performance monitoring and statistics

## Testing Strategy

- Unit tests for adaptive selection algorithms
- Integration tests with evaluators (e.g., [OLMES](https://github.com/allenai/olmes))
- Performance benchmarks comparing adaptive vs. static evaluation
- Validation that evaluation doesn't significantly impact training throughput
- Tests for difficulty estimation accuracy

## Open Questions

1. Should difficulty parameters be pre-computed or estimated on-the-fly?
2. How to handle distributed training scenarios (evaluation on rank 0 vs. all ranks)?
3. Should we support both IRT-based and simpler confidence-based selection?
4. How to balance evaluation frequency with training efficiency?
5. Integration with existing `DownstreamEvaluatorCallbackConfig`?

## Integration with Existing Code

This feature would build on the existing `EvaluatorCallback` and `DownstreamEvaluatorCallbackConfig` infrastructure, making it a natural extension of current capabilities.

Key files that would need modification or extension:

- `src/olmo_core/train/callbacks/evaluator_callback.py` - Extend existing callback
- `src/olmo_core/train/callbacks/__init__.py` - Export new callback
- New file: `src/olmo_core/train/callbacks/fluid_evaluator_callback.py` - Main implementation
- Can integrate with the `olmo_eval` package (evaluators backed by [OLMES](https://github.com/allenai/olmes)) as an optional external integration

**Comment by ada-ggf25** [comment]
Hi! I’ve been thinking about this feature for a few weeks now, before proposing it, and used that time to read the Fluid and Signal & Noise papers and get familiar with the existing eval callbacks in this repo.

If this seems like a reasonable appropriate direction for OLMo-core, I’d be happy to start on a Phase-1 MVP in my fork and open a PR once there’s something reviewable. Very open to suggestions on scope, design, or better ways to integrate this with the current callback / evaluator setup, and happy to adjust based on your comments.

**Comment by tyler-romero** [comment]
Hi Guilherme,

Thanks for your proposal

I asked Valentin Hofmann about this and he pointed out that Fluid LM Benchmarking selects questions within a task adaptively, not entire tasks. So I think some other research might be needed to come up with a principled task selection strategy.

A few other thoughts:
1. In-loop evals are very cheap, they take a few seconds total and run about every 1000 training steps (when we checkpoint) so they are very cheap from an amortized perspective - efficiency gains here might not be worth much. Its the downstream (generative) evals that are more expensive.
2. The eval tasks we do use are selected before training even begins to improve signal-to-noise (see sec 3.3 here https://www.datocms-assets.com/64837/1763662397-1763646865-olmo_3_technical_report-1.pdf#page=9.66)
3. We try to select a set of tasks that are informative across the lifetime of a pretraining run. For example, MMLU is pretty flat early on, experiences a bit of a takeoff, and then continues to rise throughout training. Its also useful to us to have a record of all in-loop evals at every checkpoint, rather than a subset of the evals for a dynamically selected subset of checkpoints (eg for comparing against other training runs).

**Comment by davidheineman** [comment]
Hi! Just chiming in as well!

Earlier this year we had a discussion of whether to use adaptive testing for in-loop eval -- the problem is that for our large runs we will map instances across hosts, then collect the eval results. Since adaptive testing is sequential, we would be eval'ing on less instances but can't take advantage of parallel compute. So, we would not really be improving eval wall time. 

I think a better place for this work would be in the generative evals in our downstream setup (in OLMES), as Tyler mentioned.

**Comment by ada-ggf25** [comment]
Hi! Thanks a lot for the detailed feedback, Tyler and David!

- You’re right that Fluid is about adaptive item selection within a task, not “adaptive task selection.” My original framing mixed those together.
- For OLMo-core in-loop eval, the combination of (a) eval being cheap/amortised, (b) the desire to keep a *complete* record at every checkpoint, and (c) the distributed setup that maps instances across hosts, makes “sequential” adaptive testing a poor fit, it wouldn’t improve wall time much and would reduce comparability/coverage.
- Also, you already select the in-loop eval task set ahead of time for signal-to-noise across the training run, so changing the task suite dynamically is not the right direction here.

Following that guidance, I opened a new issue in OLMES focused on adaptive item selection for expensive downstream/generative evaluations, where the per-item cost is dominated by external judge/API calls and reducing item count actually matters:

- https://github.com/allenai/olmes/issues/34

Key points of the new proposal:
- **No adaptive task selection** (suite remains fixed). Adaptivity is strictly within-task and opt-in dev-mode.
- **Static/full-coverage evaluation remains the default** for official reporting and cross-model comparability.
- **The MVP is batch-adaptive:** selection happens between batches, but each batch can still be sharded/parallelised across workers/hosts, to keep parallelism where it matters.

**Comment by tyler-romero** [comment]
Thanks for following up, going to close this issue for now
